### ISSUE TITLE
[query] or_error generates unfriendly stack trace

### ISSUE THREAD

**Issue by akotlar** [hail-is/hail#9226]
Found during ndarray debugging, but generally applies:

Sumary of issue:
```sh
In [21]: hl.eval(hl_nd_array_t[5,0:1000])                                                                                                                                                                                                                                                                 
---------------------------------------------------------------------------
FatalError                                Traceback (most recent call last)
<ipython-input-21-f94ec164ffe0> in <module>
----> 1 hl.eval(t[5,0:1000])

<decorator-gen-716> in eval(expression)

~/projects/hail/hail/python/hail/typecheck/check.py in wrapper(__original_func, *args, **kwargs)
    612     def wrapper(__original_func, *args, **kwargs):
    613         args_, kwargs_ = check_all(__original_func, args, kwargs, checkers, is_method=is_method)
--> 614         return __original_func(*args_, **kwargs_)
    615 
    616     return wrapper

~/projects/hail/hail/python/hail/expr/expressions/expression_utils.py in eval(expression)
    188     Any
    189     """
--> 190     return eval_timed(expression)[0]
    191 
    192 

<decorator-gen-714> in eval_timed(expression)

~/projects/hail/hail/python/hail/typecheck/check.py in wrapper(__original_func, *args, **kwargs)
    612     def wrapper(__original_func, *args, **kwargs):
    613         args_, kwargs_ = check_all(__original_func, args, kwargs, checkers, is_method=is_method)
--> 614         return __original_func(*args_, **kwargs_)
    615 
    616     return wrapper

~/projects/hail/hail/python/hail/expr/expressions/expression_utils.py in eval_timed(expression)
    154         if ir_type != expression.dtype:
    155             raise ExpressionException(f'Expression type and IR type differed: \n{ir_type}\n vs \n{expression_type}')
--> 156         return Env.backend().execute(expression._ir, True)
    157     else:
    158         uid = Env.get_uid()

~/projects/hail/hail/python/hail/backend/spark_backend.py in execute(self, ir, timed)
    294         jir = self._to_java_value_ir(ir)
    295         # print(self._hail_package.expr.ir.Pretty.apply(jir, True, -1))
--> 296         result = json.loads(self._jhc.backend().executeJSON(jir))
    297         value = ir.typ._from_json(result['value'])
    298         timings = result['timings']

~/miniconda3/lib/python3.7/site-packages/py4j/java_gateway.py in __call__(self, *args)
   1255         answer = self.gateway_client.send_command(command)
   1256         return_value = get_return_value(
-> 1257             answer, self.gateway_client, self.target_id, self.name)
   1258 
   1259         for temp_arg in temp_args:

~/projects/hail/hail/python/hail/backend/spark_backend.py in deco(*args, **kwargs)
     39             raise FatalError('%s\n\nJava stack trace:\n%s\n'
     40                              'Hail version: %s\n'
---> 41                              'Error summary: %s' % (deepest, full, hail.__version__, deepest)) from None
     42         except pyspark.sql.utils.CapturedException as e:
     43             raise FatalError('%s\n\nJava stack trace:\n%s\n'

FatalError: HailException: Index 5 is out of bounds for axis 0 with size 2

Java stack trace:
is.hail.utils.HailException: Index 5 is out of bounds for axis 0 with size 2
	at __C889Compiled.apply(Unknown Source)
	at is.hail.expr.ir.CompileAndEvaluate$$anonfun$1.apply$mcJ$sp(CompileAndEvaluate.scala:41)
	at is.hail.expr.ir.CompileAndEvaluate$$anonfun$1.apply(CompileAndEvaluate.scala:41)
	at is.hail.expr.ir.CompileAndEvaluate$$anonfun$1.apply(CompileAndEvaluate.scala:41)
	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:69)
	at is.hail.expr.ir.CompileAndEvaluate$._apply(CompileAndEvaluate.scala:41)
	at is.hail.backend.spark.SparkBackend.is$hail$backend$spark$SparkBackend$$_execute(SparkBackend.scala:318)
	at is.hail.backend.spark.SparkBackend$$anonfun$execute$1.apply(SparkBackend.scala:305)
	at is.hail.backend.spark.SparkBackend$$anonfun$execute$1.apply(SparkBackend.scala:304)
	at is.hail.expr.ir.ExecuteContext$$anonfun$scoped$1.apply(ExecuteContext.scala:20)
	at is.hail.expr.ir.ExecuteContext$$anonfun$scoped$1.apply(ExecuteContext.scala:18)
	at is.hail.utils.package$.using(package.scala:609)
	at is.hail.annotations.Region$.scoped(Region.scala:18)
	at is.hail.expr.ir.ExecuteContext$.scoped(ExecuteContext.scala:18)
	at is.hail.backend.spark.SparkBackend.withExecuteContext(SparkBackend.scala:230)
	at is.hail.backend.spark.SparkBackend.execute(SparkBackend.scala:304)
	at is.hail.backend.spark.SparkBackend.executeJSON(SparkBackend.scala:324)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)



Hail version: 0.2.53-e245ebe1d9a6
Error summary: HailException: Index 5 is out of bounds for axis 0 with size 2
```

Numpy:
```sh
In [22]: hl.eval(ndarray_t[5,0:1000])     
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-22-0643c9661056> in <module>
----> 1 hl.eval(ndarray_t[5,0:1000])

IndexError: index 5 is out of bounds for axis 0 with size 2
```

The stack trace is uninformative, and the typical Hail user will not want to see it.

**Comment by johnc1231** [comment]
I think one answer to this would be to just cut out the Java parts and show just the summary. But I also wrote up some further thoughts for how to improve here: https://dev.hail.is/t/better-python-error-messages-idea/201

**Comment by akotlar** [comment]
Thanks John. Omitting the Java parts for users by default seems like a low-effort way of handling this, but I'm open to other proposals!

**Comment by johnc1231** [comment]
I want to get it to not only omit the Java, but also show the desirable python stack trace that points to the source of the error.

**Comment by tpoterba** [comment]
This is on the radar, so I vote for closing this issue. We're using issues for bugs, and this seems more in the realm of a feature request / usability improvement.

**Comment by johnc1231** [comment]
Fixed by #9398
